src/libraries/System.Net.NameResolution/src/System/Net/Dns.cs

```diff
@@ -466,7 +524,44 @@ private static Task GetHostEntryOrAddressesCoreAsync(string hostName, bool justR
             if (NameResolutionPal.SupportsGetAddrInfoAsync && ipAddress is null)
             {
                 ValidateHostName(hostName);
-                return NameResolutionPal.GetAddrInfoAsync(hostName, justAddresses);
+
+                if (NameResolutionTelemetry.IsEnabled)
+                {
+                    ValueStopwatch stopwatch = NameResolutionTelemetry.Log.ResolutionStart(hostName);
+
+                    Task coreTask;
+                    try
+                    {
+                        coreTask = NameResolutionPal.GetAddrInfoAsync(hostName, justAddresses);
+                    }
+                    catch when (LogFailure(hostName, stopwatch))
+                    {
+                        Debug.Fail("LogFailure should return false");
+                        throw;
+                    }
+
+                    coreTask.ContinueWith(
+                        (task, state) =>
+                        {
+                            var tuple = (Tuple<string, ValueStopwatch>)state!;
+                            NameResolutionTelemetry.Log.AfterResolution(
+                                hostNameOrAddress: tuple.Item1,
+                                stopwatch: tuple.Item2,
+                                successful: task.IsCompletedSuccessfully);
+                        },
+                        state: Tuple.Create(hostName, stopwatch),
+                        cancellationToken: default,
+                        TaskContinuationOptions.ExecuteSynchronously,
+                        TaskScheduler.Default);
+
+                    // coreTask is not actually a base Task, but Task<IPHostEntry> / Task<IPAddress[]>
+                    // We have to return it and not the continuation
+                    return coreTask;
+                }
+                else
+                {
+                    return NameResolutionPal.GetAddrInfoAsync(hostName, justAddresses);
+                }
```

Comment:
Alternatively, the PAL itself [could be instrumented](https://gist.github.com/MihaZupan/882eac6ec582ef07dc74fba014011102/revisions) at the cost of increasing the state object by the size of ValueStopwatch (8 bytes)